### ISSUE TITLE
median with float64, lot of nans, and dim=1 returns random results in CUDA

### ISSUE THREAD

**Issue by Heerozh** [pytorch/pytorch#38018]
## 🐛 Bug

Not sure if it is related to #30818.

```
x = torch.full((20, 3000), np.nan, dtype=torch.float64).cuda()
x[:, :280] = 1.1
torch.median(x, dim=1).values
```

tensor([ 1.1000, 29.4100, 15.5200, 17.0700, 18.7800, 20.6600, 22.7300, 25.0000,
        27.5000, 27.0000, 15.5200, 17.0700, 18.7800, 20.6600, 22.7300, 25.0000,
        27.5000, 27.0000, 15.5600, 20.5500], device='cuda:0',
       dtype=torch.float64)

```
torch.median(x.to(torch.float32), dim=1).values
```
tensor([-1.5882e-23,  1.8875e+00, -2.0000e+00,  2.9595e+00,  2.7350e+00,
         2.7667e+00,  1.0842e-19,  2.7934e+00, -2.0000e+00,  2.8228e+00,
         2.0000e+00,  2.8552e+00,  2.8906e+00,  2.9297e+00,  2.9219e+00,
         2.8906e+00,  2.9297e+00,  2.9219e+00,  2.9219e+00,  2.9219e+00],
       device='cuda:0')

```
torch.median(x.cpu(), dim=1).values
```
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       dtype=torch.float64)

## Environment

PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 445.87
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] libblas                   3.8.0                    14_mkl    conda-forge
[conda] liblapack                 3.8.0                    14_mkl    conda-forge
[conda] mkl                       2019.4                      245
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.0.15           py37h14836fe_0
[conda] mkl_random                1.1.0            py37h675688f_0
[conda] numpy                     1.18.1           py37h93ca92e_0
[conda] numpy-base                1.18.1           py37hc3f5095_1
[conda] pytorch                   1.5.0           py3.7_cuda102_cudnn7_0    pytorch
[conda] torchvision               0.6.0                py37_cu102    pytorch

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @ngimel

**Comment by ngimel** [comment]
This is a silent wrong answer, bumping priority.

**Comment by emcastillo** [comment]
I have been able to create a small reproduced for this

```
import torch
import numpy as np

x = torch.full((2, 4), np.nan, dtype=torch.float64).cuda()
x[:, :1] = 1.1
r = torch.median(x, dim=1).values
print(r)
r = torch.median(x.cpu(), dim=1).values
print(r)
```

Should make debug easier

**Comment by xwang233** [comment]
Well, I wasn't able to exactly reproduce the original poster's code, but I found something more interesting. 

```python
import torch
import numpy as np

print(torch.__version__)
print(np.__version__)

device = 'cuda'

x = torch.tensor([
    [np.nan, 1.1, np.nan],
    [np.nan, 1.1, np.nan]
], dtype=torch.float, device=device)

# print(x)

a = torch.median(x, dim=1)
print(a)
```
gives
```
1.6.0a0+503be4e
1.17.3
torch.return_types.median(
values=tensor([0., 0.], device='cuda:0'),
indices=tensor([0, 0], device='cuda:0'))
```

However,
```python
import torch
import numpy as np

print(torch.__version__)
print(np.__version__)

device = 'cuda'

x = torch.tensor([
    [np.nan, 1.1, np.nan],
    [np.nan, 1.1, np.nan]
], dtype=torch.float, device=device)

print(x)    # Note: a print(x) here!!

a = torch.median(x, dim=1)
print(a)
```
gives
```
1.6.0a0+503be4e
1.17.3
tensor([[   nan, 1.1000,    nan],
        [   nan, 1.1000,    nan]], device='cuda:0')
torch.return_types.median(
values=tensor([-1.0842e-19,  1.8875e+00], device='cuda:0'),
indices=tensor([4607632778870128640, 4607632778870128640], device='cuda:0'))
```
Note that you can get this in C++
```c++
int main(){
    long a = 4607632778870128640;
    printf("%f\n", *(double*)&a);   // 1.100000
}
```
which is very interesting. 

My environment
```
$ python -c 'from torch.utils.collect_env import main; main()'
Collecting environment information...
PyTorch version: 1.6.0a0+503be4e
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Fedora 31 (Workstation Edition)
GCC version: (GCC) 8.3.1 20190223 (Red Hat 8.3.1-2)
CMake version: version 3.15.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2070 SUPER
Nvidia driver version: 440.31
cuDNN version: /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.17.3
[pip3] torch==1.6.0a0+503be4e
[pip3] torchvision==0.7.0a0+f9ef235
[conda] Could not collect
```

**Comment by emcastillo** [comment]
in `eq_with_nan` having `v` and `kValue` both `nan` is returning false when it should be true.
The implementation for double is using `__double_as_longlong` and comparing their bits,
but the bits values obtained for both are different!
` 9221120237041090560` for v
`9223372036854775807` for kValue

**Comment by ngimel** [comment]
#16762 can be related (it's topK though, not media/Kth value).
